### PR TITLE
tbls: library for BLS12-381 abstraction

### PR DESCRIPTION
**Pull request by gsora**
This PR adds a subpackage in `tbls` called `v2`, which is a new abstraction for BLS12-381 cryptography.

It contains two concrete implementations — one using [Herumi](https://github.com/herumi/bls-eth-go-binary) and another based on [Kryptology](https://github.com/coinbase/kryptology), added for backwards compatibility while we work out the migration.

A complete test suite is provided, alongside a fuzzing target to test cross-compatibility between implementation.

Preliminary tests show substantial performance improvements. 

category: refactor
ticket: https://github.com/ObolNetwork/charon/issues/1658

### CODE REVIEW

**Comment by gsora** [comment]
Fixing linting issues as we speak.

File: go.mod [hunk 1]
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/ferranbt/fastssz v0.1.2
 	github.com/golang/snappy v0.0.4
 	github.com/gorilla/mux v1.8.0
+	github.com/herumi/bls-eth-go-binary v1.28.1
```
Comment (corverroos) [inline go.mod:13]: might need to remove any references of "CGO_ENABLED=0"

File: tbls/v2/herumi/herumi.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	"sync"
+
+	"github.com/herumi/bls-eth-go-binary/bls"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/z"
+	v2 "github.com/obolnetwork/charon/tbls/v2"
+)
+
+var initializationOnce = sync.Once{}
```
Comment (corverroos) [inline tbls/v2/herumi/herumi.go:30]: nit, short variables FTW: `initOnce`

File: tbls/v2/herumi/herumi.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+// Herumi is an Implementation with Herumi-specific inner logic.
+type Herumi struct{}
+
+func (Herumi) GenerateSecretKey() (v2.PrivateKey, error) {
+	var p bls.SecretKey
+	p.SetByCSPRNG()
+
+	// Commenting here once, this syntax will appear often:
+	// here I'm converting ret to a pointer to instance of v2.PrivateKey, which is
```
Comment (corverroos) [inline tbls/v2/herumi/herumi.go:58]: no "ret" here...

File: tbls/v2/tbls.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	Verify(compressedPublicKey PublicKey, data []byte, signature Signature) error
+
+	// Sign signs data with the provided private key, and returns the resulting signature.
+	// This function works on both shares of private keys, and complete private keys.
+	Sign(privateKey PrivateKey, data []byte) (Signature, error)
+}
+
+// SetImplementation sets newImpl as the package backing implementation.
+func SetImplementation(newImpl Implementation) {
+	implLock.Lock()
```
Comment (corverroos) [inline tbls/v2/tbls.go:66]: now that I see it, lock only makes sense if we also use it when accessing/reading the global variable in funcs below. Lets rather remove the lock and add a comment: `this isn't threadsafe and should be called on startup before using any functions below`...?

**Review by corverroos — APPROVED** [review]

File: tbls/v2/tbls.go (hunk 4 above)
Comment (corverroos) [reply tbls/v2/tbls.go:66]: sure no problem

**Pull request merged**